### DISCUSSION TITLE
Welcome to order-matcher Discussions!

### DISCUSSION THREAD

**Discussion by ArjunVachhani** [ArjunVachhani/order-matcher#62] [Announcements]
## 👋 Welcome!
  We’re using Discussions as a place to connect with other members of our community. We hope that you:
  * Ask questions you’re wondering about.
  * Share ideas.
  * Engage with other community members.
  * Welcome others and are open-minded. Remember that this is a community we
  build together 💪.

  To get started, comment below with an introduction of yourself and tell us about what you do with this community.

**Comment by smn333** [comment]
Hi,
I'm an engineering student and I was trying to understand the code, 
Is it that each MatchingEngine instance has to be created for each stock right? like if we want order matcher for 10 companies like microsoft and google etc., so each company we create a new (new MatchingEngine ()) instance . 
Is this correct? - Also is this the efficient way to do ? 
If what I understood is correct then user should track the multiple MatchEngine instances? Is that correct?

**Comment by ArjunVachhani** [comment]
Hi @smn333 

> Is it that each MatchingEngine instance has to be created for each stock right? like if we want order matcher for 10 companies like microsoft and google etc., so each company we create a new (new MatchingEngine ()) instance .

Yes, for each company you need to create a new instance of `MatchineEngine`

> Is this correct? - Also is this the efficient way to do ?

Yes, `MatchineEngine` is light weight class. You need to create new instance of  `MatchineEngine` class for each company/trading pair.

> If what I understood is correct then user should track the multiple MatchEngine instances? Is that correct?

Not sure what you meant by "user should track the multiple MatchEngine". But you can pass TradeListener in each instance to listen for events and then you can then broadcast events to relevant audience.

**Reply by smn333** [reply]
Hi @ArjunVachhani ,

Thank you so much for the quick response.
So, if we create two Matching Engine instances for two different company/trading pair - suppose A and B.
And there is a buy/bid for both A and B at exact same point of time, how will it be processed?

I mean it is sequentially processed right? - and we need to handle it parallely right? or are these handled in the deployment scenario and code remains same?

I am trying to understand is this how in real world matching engine works in a stock exchange -- in your code - I ideally find functionality of an exachange done, so was trying to understand is it that - how we handle matching engine objects is what forms an exchange..

Thanks,
Subramanya

**Reply by ArjunVachhani** [reply]
Hi @smn333 

> So, if we create two Matching Engine instances for two different company/trading pair - suppose A and B.
> And there is a buy/bid for both A and B at exact same point of time, how will it be processed?

For each company sequence of order should be preserved (price time priority). If both order of same company then it should be processed sequentially. In your example you can process bids of both company parallelly as they are trading different instrument.

**Reply by smn333** [reply]
HI @ArjunVachhani , 
Got it, Thanks for the response.
One more doubt, you qualified 1million transactions per second, but how was this tested? did you use any load testing tools or can you guide how to test if we clone your repo?
Thanks,
Subramanya

**Reply by ArjunVachhani** [reply]
Sure, code is under performance-tesstint-7.0 branch.

Here is a file that measures it.
https://github.com/ArjunVachhani/order-matcher/blob/performance-testing-7.0/OrderMatcher/OrderMatcher.Performance/LoadRunner.cs
